### ISSUE TITLE
GraphQL client: "endpoint" parameter in "@GraphQLClientApi" has precedence over "application.properties"

### ISSUE THREAD

**Issue by jmini** [quarkusio/quarkus#44334]
From comment https://github.com/quarkusio/quarkus/pull/44293#discussion_r1829350526:

When testing the GraphQL client I discovered that the endpoint defined in the annotation:

```
@GraphQLClientApi(configKey = "gitlab", endpoint = "https://gitlab.com/api/graphql")
```

Has precedence over what is defined in the application properties:

```
quarkus.smallrye-graphql-client.gitlab.url=https://gitlab.example.org/api/graphql
```

Example commit to "move" the config property:
https://github.com/jmini/gitlab-experiments/commit/a79a36c134e61ce4abe9c8e77974fd787b3d47c7

According to @jmartisk it should be the other way around.

---

Note that the settings made in TypesafeGraphQLClientBuilder wins over what is defined in the application properties, but I think that this is correct:

```
TypesafeGraphQLClientBuilder.newBuilder()
                .endpoint("https://gitlab.example.org/api/graphql")
                .header("Authorization", "Bearer " + gitlabToken)
                .build(WorkitemClientApi.class);
```

**Comment by jmini** [comment]
@jmartisk If you point me to the location of the integration test in quarkus, I could try to add a unit test to validate my observations (and if the expectation is that "application.properties" should win, then this test should fail)

**Comment by jmini** [comment]
@jmartisk I think my test pushed here:
https://github.com/jmini/quarkus/commit/27845afe4b665f7276535b311ddd4dd4877f0325
Validates that `endpoint` in `@GraphQLClientApi` has precedence over `application.properties`

Any idea what I should investigate for the fix?

Maybe the place where the configs are merged?

**Comment by mskacelik** [comment]
https://github.com/smallrye/smallrye-graphql/pull/2223

**Comment by jmartisk** [comment]
Fixed via https://github.com/smallrye/smallrye-graphql/pull/2223, so it will be included in Quarkus 3.17 (probably)

@jmini do you want to submit a test yourself (you'd need to change your commit with the test to add a new test instead of changing an existing one)?  The fix is not in Quarkus yet, but if you want to do it now, I can start a new Quarkus branch that will serve for capturing changes related to SmallRye GraphQL 2.12, and you could send a PR to that branch. Or we can tell @mskacelik to take care of it, whatever you prefer.

**Comment by jmini** [comment]
Don't you need a release of SmallRye GraphQL `2.12` first?

I am happy to contribute my test, I will split the test as required and open a draft MR because it is easier to discuss on code changes. (the draft PR can remain open until there is a SmallRye release)

From my side the release and the quarkus update is not urgent, the work-around of removing the `endpoint` attribute in the annotation (as I did in https://github.com/jmini/gitlab-experiments/commit/a79a36c134e61ce4abe9c8e77974fd787b3d47c7) is really not a blocker for me.

**Comment by jmartisk** [comment]
We usually do it the way that I maintain a custom branch of Quarkus that captures the changes related to the next smallrye-graphql upgrade, and the branch depends on a snapshot of the upcoming release version. But you can as well just send a draft PR straight to Quarkus (without actually changing the dependency version, so the test will just fail for now) and I will make sure to merge it after the smallrye-graphql upgrade.

**Comment by jmini** [comment]
> and the branch depends on a snapshot of the upcoming release version.

If you already created that branch, I can target this with my PR and then the test should be green.

**Comment by jmartisk** [comment]
That won't work like this OOTB because the Quarkus CI doesn't know how to build a snapshot of smallrye-graphql, and we don't deploy snapshots into public Maven repositories.
Let's keep it simple and just send a draft PR to Quarkus? If you locally develop it against a smallrye-graphql snapshot in a way that the test passes, it will be easy for me to simply merge it after the upgrade.

**Comment by jmini** [comment]
>  Let's keep it simple and just send a draft PR to Quarkus? If you locally develop it against a smallrye-graphql snapshot in a way that the test passes, it will be easy for me to simply merge it after the upgrade.

This was my first proposal. 😉 

> If you locally develop it against a smallrye-graphql snapshot in a way that the test passes, it will be easy for me to simply merge it after the upgrade.

I don't have that setup on my machine. But maybe you can try my branch on your machine locally. Otherwise we hope that the change in https://github.com/jmini/gitlab-experiments/commit/a79a36c134e61ce4abe9c8e77974fd787b3d47c7 is good, but we will only be sure after the smallrye release.

**Comment by jmartisk** [comment]
Ok I'll do a test run with https://github.com/jmini/quarkus/commit/26d61f65cc28b68a1c81814878eaf6640781afff and let you know

**Comment by jmartisk** [comment]
Yeah it worked, just feel free to send a draft PR with https://github.com/jmini/quarkus/commit/26d61f65cc28b68a1c81814878eaf6640781afff and say in the description that it requires an upgrade SmallRye GraphQL 2.11.1 or 2.12.0, and I'll take care of it

**Comment by jmini** [comment]
See https://github.com/quarkusio/quarkus/pull/44441
